packages/router/router.ts

```diff
@@ -3093,11 +3219,91 @@ function shouldRevalidateLoader(
   return arg.defaultShouldRevalidate;
 }
 
+/**
+ * Execute route.lazy() methods to lazily load route modules (loader, action,
+ * shouldRevalidate) and update the routeManifest in place which shares objects
+ * with dataRoutes so those get updated as well.
+ */
+async function loadLazyRouteModule(
+  route: AgnosticDataRouteObject,
+  hasErrorBoundary: HasErrorBoundaryFunction,
+  manifest: RouteManifest
+) {
+  if (!route.lazy) {
+    return;
+  }
+
+  let lazyRoute = await route.lazy();
+
+  // If the lazy route function has already been executed and removed from
+  // the route object by another call while we were waiting for the promise
+  // to resolve then we don't want to resolve the same route again.
+  if (!route.lazy) {
+    return;
```

Comment:
Just calling out for review, but right now it's "first resolved promise wins" in our code so _theoretically_ non-deterministic and subject to race conditions.  However I think in most (all?) cases `import()` semantics will handle this for us so it should be a non-issue in practice.  Plus, `lazy()` shouldn't ever not be returning different info since routes are immutable once loaded, so hard to think of a use-case where caching actually fixes anything?

Reply:
I added this for a couple of reasons:
- It avoids all the work and validation messages below from being repeated.
- In the rare case where someone has configured their `lazy` function to return different results, the failure case is at least a little more predictable in that it will never merge results from multiple calls into a single object, instead leaving the first result intact.

It seemed like an easy win given the tiny amount of code needed, but I could live without it.

Reply:
This API is 👍